scylla/tests/retries.rs

```diff
@@ -0,0 +1,284 @@
+    let real2_addr = SocketAddr::from_str(real2_uri.as_str()).unwrap();
+    let proxy2_addr = SocketAddr::from_str(proxy2_uri.as_str()).unwrap();
+    let real3_addr = SocketAddr::from_str(real3_uri.as_str()).unwrap();
+    let proxy3_addr = SocketAddr::from_str(proxy3_uri.as_str()).unwrap();
+
+    let proxy = Proxy::builder()
+        .with_node(
+            Node::builder()
+                .real_address(real1_addr)
+                .proxy_address(proxy1_addr)
+                .shard_awareness(ShardAwareness::QueryNode)
+                .build(),
+        )
+        .with_node(
+            Node::builder()
+                .real_address(real2_addr)
+                .proxy_address(proxy2_addr)
+                .shard_awareness(ShardAwareness::QueryNode)
+                .build(),
+        )
+        .with_node(
+            Node::builder()
+                .real_address(real3_addr)
+                .proxy_address(proxy3_addr)
+                .shard_awareness(ShardAwareness::QueryNode)
+                .build(),
+        )
+        .build();
+
+    let running_proxy = proxy.run().with_current_subscriber().await.unwrap();
+
+    let mut translation_map = HashMap::new();
+    for (real_addr, proxy_addr) in [
+        (real1_addr, proxy1_addr),
+        (real2_addr, proxy2_addr),
+        (real3_addr, proxy3_addr),
+    ] {
+        translation_map.insert(real_addr, proxy_addr);
+    }
+
+    let running_proxy = test(
+        [proxy1_uri, proxy2_uri, proxy3_uri],
+        translation_map,
+        running_proxy,
+    )
+    .await;
+
+    running_proxy.finish().await
+}
+
+#[tokio::test]
+#[ntest::timeout(5000)]
+async fn speculative_execution_is_fired() {
+    const TIMEOUT_PER_REQUEST: Duration = Duration::from_millis(200);
+    let res = test_with_3_node_cluster(217, |proxy_uris, translation_map, mut running_proxy| async move {
+        // DB preparation phase
+        let session: Session = SessionBuilder::new()
+            .known_node(proxy_uris[0].as_str())
+            .speculative_execution(Arc::new(SimpleSpeculativeExecutionPolicy {
+                max_retry_count: 2,
+                retry_interval: Duration::from_millis(10),
+            }))
+            .retry_policy(Box::new(FallthroughRetryPolicy))
+            .address_translator(Arc::new(translation_map))
+            .build()
+            .await
+            .unwrap();
+
+        session
+            .query("DROP KEYSPACE IF EXISTS ks2", &[])
+            .await
+            .unwrap();
+        session.query("CREATE KEYSPACE IF NOT EXISTS ks2 WITH REPLICATION = {'class' : 'SimpleStrategy', 'replication_factor' : 3}", &[]).await.unwrap();
+        session
+            .query("CREATE TABLE ks2.t (a int primary key)", &[])
+            .await
+            .unwrap();
+
+        let mut q = Query::from("INSERT INTO ks2.t (a) VALUES (?)");
+        q.set_is_idempotent(true); // this is to allow speculative execution to fire
+
+        let delay_rule = RequestRule(
```

Comment:
It's called `delay_rule`, but it drops requests instead of delaying them.

Reply:
Yes, please. I guess that "infinite delay" == "drop" in some sense, but it's still confusing. When I saw the name `delay_rule` but saw that it uses `drop_frame()` instead of the other, existing method called `delay()`, my first thought was: is it a mistake?

Reply:
By the way, there is another `delay_rule` in the code which does not drop or delay but rather forges a request error. That one looks even weirder.